consensus/src/error.rs

```diff
@@ -0,0 +1,79 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub struct DbError {
+    #[from]
+    inner: anyhow::Error,
+}
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub struct StateSyncError {
+    #[from]
+    inner: anyhow::Error,
+}
+
+impl From<executor_types::Error> for StateSyncError {
+    fn from(e: executor_types::Error) -> Self {
+        StateSyncError { inner: e.into() }
+    }
+}
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub struct MempoolError {
+    #[from]
+    inner: anyhow::Error,
+}
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+pub struct VerifyError {
+    #[from]
+    inner: anyhow::Error,
+}
+
+pub fn error_kind(e: &anyhow::Error) -> &'static str {
+    if e.downcast_ref::<executor_types::Error>().is_some() {
+        return "Execution";
+    }
+    if let Some(e) = e.downcast_ref::<StateSyncError>() {
+        if e.inner.downcast_ref::<executor_types::Error>().is_some() {
+            return "Execution";
+        }
+        return "StateSync";
+    }
+    if e.downcast_ref::<MempoolError>().is_some() {
+        return "Mempool";
+    }
+    if e.downcast_ref::<DbError>().is_some() {
+        return "ConsensusDb";
+    }
+    if e.downcast_ref::<safety_rules::Error>().is_some() {
+        return "SafetyRules";
+    }
+    if e.downcast_ref::<VerifyError>().is_some() {
+        return "VerifyError";
+    }
+    "InternalError"
+}
```

Comment:
shouldn't this be obvious by the error type? why not just match on the error type? why not use an enum instead? and just have a ConsensusError or a generic Error?